src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java

```diff
@@ -2904,19 +2916,71 @@ private SocketFactory getSocketFactory() throws IOException {
         return socketFactory;
     }
 
-    // This method contains the old logic of connecting to
-    // a socket of one of the IPs corresponding to a given host name.
-    // In the old code below, the logic around 0 timeout has been removed as
-    // 0 timeout is not allowed. The code has been re-factored so that the logic
-    // is common for hostName or InetAddress.
-    private Socket getDefaultSocket(String hostName, int portNumber, int timeoutInMilliSeconds) throws IOException {
-        // Open the socket, with or without a timeout, throwing an UnknownHostException
-        // if there is a failure to resolve the host name to an InetSocketAddress.
-        //
+    /**
+     * This method makes a connected socket given a hostname or IP address.
+     * 
+     * @param hostName
+     *        Hostname or IP Address
+     * @param portNumber
+     *        Port number
+     * @param timeoutInMilliSeconds
+     *        Timeout
+     * @param IPAddressPreference
+     *        IPv4First (default), IPv6First or UsePlatformDefault
+     * @return Connected Socket
+     * @throws IOException
+     * @throws SQLServerException
+     */
+    private Socket getSocketByIPvPreference(String hostName, int portNumber, int timeoutInMilliSeconds,
+            String IPAddressPreference) throws IOException, SQLServerException {
+        InetSocketAddress addr = null;
+        InetAddress addresses[] = InetAddress.getAllByName(hostName);
+        Queue<InetAddress> addrq = null;
+
+        if (IPAddressPreference.equals("IPv6First")) {
+            // Try to connect to IPv6 Addresses
+            addrq = getIPv6Addresses(addresses);
+            while (addrq.peek() != null) {
```

Comment:
we should do some fine or finest level logging here

Reply:
I think that can potentially result in a large amount of logging for systems with many IP addresses resolving. The cases of no ip addresses resolving will be caught by the log entry below.